### PR TITLE
Issue 1578: Rejection reason (RJM)

### LINKED ISSUE #1578: In registrar admin, allow an analyst to set a "reason for rejection" and have the registrar send an email based upon the reason
### Issue description

As an analyst
I want to tell a requestor why their domain name was rejected in the same place I reject it
so that...
* we track rejection reasons as a category in the registrar
* users don't have to receive two emails about their request being rejected
* I don't have to write a separate email 

In the past, people whose request had been rejected got an automated email from the registrar simply stating the request was rejected _and_ a bespoke, analyst-crafted email that included the rejection reason. We've now [summarized the rejection reasons](https://github.com/cisagov/manage.get.gov/issues?q=is%3Aissue+is%3Aopen+reject+author%3Agabydisarli+) into 6 different circumstances. 

In the registrar admin, after setting the rejection status, an analyst should be able to select the "rejected reason". Once saved, the requestor should be sent the email template that corresponds to the rejected reason.

[Rejection email templates are in our Google Drive](https://drive.google.com/drive/folders/1gkUyKEad42GiwzN7carJtMAw1RHyG8KC)

### Acceptance criteria

- [ ] After setting a rejection status in the domain request, an analyst can select from the 6 rejection reasons (ideally without having to save the reject status first before viewing the reasons).
- [ ] After an analyst saves the domain request, the requestor receives a [reject-reason-dependent email](https://drive.google.com/drive/folders/1gkUyKEad42GiwzN7carJtMAw1RHyG8KC). 
- [ ] Analyst is able to **filter** by Rejection Reason code

### Additional context

When Status = Rejected, another field should be shown, "Reason for rejection". Mocked example:

image

### Links to other issues

🔄 Relates to #608

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #1578 

## Changes

- Add rejection reason to model
- Catch rejection reason is None when transition to rejected in DjA
- Hide/show rejection reason field through JS
- Clear rejection reason when transitioning from rejected (in model)
- Different emails for different rejection reasons
- Added rejection_reason filter on domain requests table (but did not add a column in the UI)
- - Additional AC: 'other' rejection reason that triggers a generic email

## Context and setup for reviewers

Note the code to handle the UI when you change the selection and navigate forward/backward using browser buttons without saving, then coming back to the form. Make sure you test these scenarios by changing from rejected, clicking on a 'eye' somewhere, then clicking 'back. Likewise, change to rejected and do the same.

Edit the submitter and add your email, then reject/unreject trying all 7 reasons to test all 7 emails.

Note for designers: I'm not in love with the long rejection reasons names in the filter. If you agree, we can increase the scope of this ticket or make another small one to have long and short reason names (like orgs).

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [x] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Screenshot 2024-02-22 at 8 21 41 PM

Screenshot 2024-02-22 at 8 21 46 PM

Screenshot 2024-02-22 at 8 25 38 PM

Screenshot 2024-02-22 at 8 25 48 PM

Screenshot 2024-02-22 at 8 25 59 PM

Screenshot 2024-02-22 at 8 26 07 PM

Screenshot 2024-02-22 at 8 26 13 PM

Screenshot 2024-02-22 at 8 26 24 PM

Screenshot 2024-02-22 at 8 26 33 PM

### CODE REVIEW

File: src/registrar/assets/js/get-gov-admin.js [hunk 1]
```diff
@@ -339,3 +339,46 @@ function enableRelatedWidgetButtons(changeLink, deleteLink, viewLink, elementPk,
...
+ * status select amd to show/hide the rejection reason
+*/
+(function (){
+    let rejectionReasonFormGroup = document.querySelector('.field-rejection_reason')
+
+    if (rejectionReasonFormGroup) {
+        let statusSelect = document.getElementById('id_status')
+
+        // Initial handling of rejectionReasonFormGroup display
+        if (statusSelect.value != 'rejected')
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov-admin.js:353]: Might wish to wrap this in if(statusSelect)....you know, for the sake of consistency and rigor

File: src/registrar/assets/js/get-gov-admin.js [hunk 2]
```diff
@@ -339,3 +339,46 @@ function enableRelatedWidgetButtons(changeLink, deleteLink, viewLink, elementPk,
...
+            } else {
+                rejectionReasonFormGroup.style.display = 'none';
+                sessionStorage.setItem('hideRejectionReason', 'true');
+            }
+        });
+    }
+
+    // Listen to Back/Forward button navigation and handle rejectionReasonFormGroup display based on session storage
+
+    // When you navigate using forward/back after changing status but not saving, when you land back on the DA page the
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov-admin.js:370]: Very nice consideration ^__^

File: src/registrar/templates/emails/status_change_rejected.txt [hunk 3]
```diff
@@ -8,9 +8,65 @@ REQUEST RECEIVED ON: {{ application.submission_date|date }}
 STATUS: Rejected
 
 ----------------------------------------------------------------
+{% if application.rejection_reason != 'other' %}
```
Comment (CocoByte) [inline src/registrar/templates/emails/status_change_rejected.txt:11]: A little confused about why we want rejection_reason == 'other' to hide this text...but I am sure it was a requirement somewhere, so I will just note my confusion and let you carry on.
NO CHANGE NEEDED

**Review by CocoByte — APPROVED** [review]
One Very Minor Javascript comment.  I am approving since it is minor and we could live without the change.  Overall, great PR to review so I'm happy I got to do this one -- nice work as always :)

**Comment by rachidatecs** [comment]
Revised email spacing and revised rejection reason definitions (copy) @Katherine-Osos

**Review by Katherine-Osos — APPROVED** [review]

**Pull request merged**